ddtrace/tracer/stats.go

```diff
@@ -0,0 +1,362 @@
+	Synthetics bool
+}
+
+type rawBucket struct {
+	start, duration uint64
+	data            map[aggregation]*rawGroupedStats
+}
+
+func newRawBucket(btime uint64, bsize int64) *rawBucket {
+	return &rawBucket{
+		start:    btime,
+		duration: uint64(bsize),
+		data:     make(map[aggregation]*rawGroupedStats),
+	}
+}
+
+func (sb *rawBucket) handleSpan(ss *aggregableSpan) {
+	gs, ok := sb.data[ss.key]
+	if !ok {
+		gs = newRawGroupedStats()
+		sb.data[ss.key] = gs
+	}
+	if ss.TopLevel {
+		gs.topLevelHits++
+	}
+	gs.hits++
+	if ss.Error != 0 {
+		gs.errors++
+	}
+	gs.duration += float64(ss.Duration)
+	// alter resolution of duration distro
+	trundur := nsTimestampToFloat(ss.Duration)
+	if ss.Error != 0 {
+		gs.errDistribution.Add(trundur)
+	} else {
+		gs.okDistribution.Add(trundur)
+	}
+}
+
+// Export transforms a RawBucket into a statsBucket, typically used
+// before communicating data to the API, as RawBucket is the internal
+// type while statsBucket is the public, shared one.
+func (sb *rawBucket) Export() statsBucket {
+	csb := statsBucket{
+		Start:    sb.start,
+		Duration: sb.duration,
+		Stats:    make([]groupedStats, len(sb.data)),
+	}
+	for k, v := range sb.data {
+		b, err := v.export(k)
+		if err != nil {
+			log.Error("Could not export stats bucket: %v.", err)
+			continue
+		}
+		csb.Stats = append(csb.Stats, b)
+	}
+	return csb
+}
+
+type rawGroupedStats struct {
+	// using float64 here to avoid the accumulation of rounding issues.
+	hits            float64
+	topLevelHits    float64
+	errors          float64
+	duration        float64
+	okDistribution  *ddsketch.DDSketch
+	errDistribution *ddsketch.DDSketch
+}
+
+func newRawGroupedStats() *rawGroupedStats {
+	const (
+		// relativeAccuracy is the value accuracy we have on the percentiles. For example, we can
+		// say that p99 is 100ms +- 1ms
+		relativeAccuracy = 0.01
+		// maxNumBins is the maximum number of bins of the ddSketch we use to store percentiles.
+		// It can affect relative accuracy, but in practice, 2048 bins is enough to have 1% relative accuracy from
+		// 80 micro second to 1 year: http://www.vldb.org/pvldb/vol12/p2195-masson.pdf
+		maxNumBins = 2048
+	)
+	okSketch, err := ddsketch.LogCollapsingLowestDenseDDSketch(relativeAccuracy, maxNumBins)
+	if err != nil {
+		log.Error("Error when creating ddsketch: %v", err)
+	}
+	errSketch, err := ddsketch.LogCollapsingLowestDenseDDSketch(relativeAccuracy, maxNumBins)
+	if err != nil {
+		log.Error("Error when creating ddsketch: %v", err)
+	}
+	return &rawGroupedStats{
+		okDistribution:  okSketch,
+		errDistribution: errSketch,
+	}
+}
+
+func (s *rawGroupedStats) export(k aggregation) (groupedStats, error) {
+	msg := s.okDistribution.ToProto()
+	okSummary, err := proto.Marshal(msg)
+	if err != nil {
+		return groupedStats{}, err
+	}
+	msg = s.errDistribution.ToProto()
+	errSummary, err := proto.Marshal(msg)
+	if err != nil {
+		return groupedStats{}, err
+	}
+	// round a float to an int, uniformly choosing
```

Comment:
You don't need to round as numbers are exact integers. No weight is involved here as all computations are made prior to sampling

Reply:
Yep. Absolutely right. Changed it.